### ISSUE TITLE
Ability to parse dev dependencies.

### ISSUE THREAD

**Issue by topherfangio** [RodneyEbanks/requirejs-plugin-bower#2]
Hello!

I just found this plugin and I'm loving it so far! I have been able to refactor a bunch of our code to use this and it is quite nice.

I was also able to use this to run our code under Karma, however, I've had to add paths/shims for all of our development dependencies that Karma uses. It would be great if this was handled by the plugin. Is it currently possible to include the top-level bower.json devDependencies along with the regular dependencies?

If not, would this be an acceptable config option? I'm happy to create a pull request if so :-)

Thanks!

**Comment by RodneyEbanks** [comment]
Just pushed an update [0153add](https://github.com/RodneyEbanks/requirejs-plugin-bower/commit/0153add0e8c9783cd0899e93c5991f0c0d01d837) which should provide you this functionality.  It is not activate by default so you have to add a line to the ...{bower: {deps:['dependencies', 'devDependencies']}}... entry in main.js.

If this doesn't resolve your issue feel free send a pull request with your resolution.

Glad you found requirejs-plugin-bower use full.

**Comment by topherfangio** [comment]
Hi Rodney!

Thanks so much for that addition!

I added a quick line-comment on the commit as it's causing me some issues, but I was able to get past that. Now, I seem to be experiencing another problem when I attempt to use this new feature. The plugin appears to be exiting before calling my callback. This works fine if I only do standard dependencies and not the devDependences.

I'm going to continue debugging it, but I have not found anything quite yet. My only initial assumption is that I'm seeing some 404 errors on some bower.json dependencies which might be causing issues.

**Comment by RodneyEbanks** [comment]
Topher one thing I found with bower install \* --save-dev option is that bower doesn't appear to install a devDependencies dependencies i.e if you install jQuery as a devDependency it is installed but not its dependencies e.g. sizzle etc.

So my plugin in parsing devDependencies will parse your root bower.json and then jQuery's bower.json and then because of jQuery's devDependecies try to process sizzle's bower.json which will be missing.  Another issue is if the module/package is missing a bower.json file (.bower.json doesn't work as the browser can't get hidden files)

Not sure if there is an easy fix as the solutions I can think of would have unintended consequences.

Will be adding a debug mode which would output a list of bower.json required but not installed, but you should see them in console.log and you can view what was loaded in require.s.contexts._.config or the bowerConfig return value in the example code.

If you could share the dependencies and devDependencies of your bower.json that would help.

Finally one last dependency for good luck.

**Comment by RodneyEbanks** [comment]
I have pushed a fix for your line comment, if you could test it on your bower.json and let me know if it works, the else statement is still there but the if was buggy, the update should correctly reflect the logic.

**Comment by topherfangio** [comment]
Hi Rodney,

Sorry for the delay! I'm out of the country at the moment and have been working quite a bit. I just now got a chance to try out your fix. Unfortunately, it still has the same behavior of executing the `onParse()` and stopping execution at that point.

Perhaps I don't understand the logic properly, but isn't the `if` statement is supposed to check if the `bowerJson` property is a string, and if so, reset the `bowerJson` to the parsed version of the string?

Should the code instead be:

``` javascript
// Ensure we have a valid bowerJson
if (bowerJson) {
  // Fixme: requirejs-plugins json Returns a javascript object inBrowser and json string inBuild
  if (typeof bowerJson !== 'object') {
    // We have a valid bowerJson, but it's a string, not an object, so we convert it
    bowerJson = JSON.parse(bowerJson);
  } else {
    // Nothing to do; we have a valid bowerJson object and should just move on;
    // (obviously this comment/else can be removed; it's here to help convey my point)
  }
} else {
  // We do NOT have a valid bowerJson object; this can happen if a dependency is in any bower.json but is not installed; most common with devDependencies
  onParse(bower.config);
}
```

Does the above logic make sense? I've added some comments to describe what I'm thinking.

A more succinct version might be:

``` javascript
// Check to make sure we actually have a bowerJson, otherwise, just call the onParse with an empty config and return
if (!bowerJson) {
  onParse(bower.config);
  return;
}

// Fixme: requirejs-plugins json Returns a javascript object inBrowser and json string inBuild
if (typeof bowerJson !== 'object') {
  // We have a valid bowerJson, but it's a string, not an object, so we convert it
  bowerJson = JSON.parse(bowerJson);
}
```

Please let me know if you have questions or if the logic is incorrect :-)

**Comment by RodneyEbanks** [comment]
No problem; I have applied your test for no file to the parseBowerFile and moved the JSON.parse to the processBowerFile which in looking at it, should have been it's proper location.  If you could give it a try and let me know if it works.

If it doesn't if you could confirm that the example is running and building in your configuration, if not if you could provide your bower.json or the dependencies/devDependencies section that would be great.

Thanks for your testing and feedback it is very helpful.

**Comment by RodneyEbanks** [comment]
Closing issue; enhancement available as an option verified in commit e31b4c81f5 authored on Feb 1.
